contents/blog/temporal-exports.md

```diff
@@ -0,0 +1,118 @@
+
+## How we built our new export system with Temporal
+
+We didn’t completely rebuild our export system; many of the pieces still worked, we just needed to modify them to work with Temporal’s scheduling and execution capabilities. This meant building a batch export abstraction layer on top of Temporal. Specifically, the abstraction layer is four models:
+
+- `BatchExportDestination`: How users configure where to export data their data.
+- `BatchExportSchedule`: Maps to a [Temporal Schedule](https://keithtenzer.com/temporal/temporal_schedules/). It stores Schedule configurations to pass to Temporal.
+- `BatchExport`: The union of a `BatchExportSchedule` with a `BatchExportDestination`. Defines a destination for exported data, and a schedule to run said exports.
+- `BatchExportRun`: A `BatchExport` for each triggered workflow. Users aren’t meant to create these, only Temporal. They mostly serve to report on individual workflow executions.
+
+Users create a `BatchExport` that includes a schedule (which validates against the `BatchExportSchedule`) and a destination (stored as a `BatchExportDestination`). Data then gets read from PostHog and processed to write to the destination.
+
+With this structure in place, we began working on improving reliability, transparency, and performance.
+
+### Improving reliability and adding retries
+
+We prioritized improving reliability. As mentioned, Temporal has support for retry and timeout logic per activity and workflow which made a big difference. What we needed to do was tune this to our system and destinations. We needed to tune four main pieces:
+
+1. **ClickHouse** - [spaced out retries](https://github.com/PostHog/posthog/pull/15558) when reconnecting to prevent failures if in use by another query, retry [connection errors](https://github.com/PostHog/posthog/pull/16574) which are likely network failures rather than outages, use [offline ClickHouse cluster](https://github.com/PostHog/posthog/pull/16470) to improve performance and allowable query length, and sacrificed some performance to [de-duplicate batches](https://github.com/PostHog/posthog/pull/16354) by adding `DISTINCT ON` clause (because it handles queries asynchronously so de-duplication isn’t guaranteed).
+2. **Postgres** - [health checks](https://github.com/PostHog/posthog/pull/16079) to ensure it is healthy before running workflows and [retry forever](https://github.com/PostHog/posthog/pull/16703) because we control everything related to those activities so failures are transient (other than bugs).
+3. **Destinations** - define which destination errors are [retriable and which aren’t](https://github.com/PostHog/posthog/pull/16814), [reset the iterator when a `JSONDecodeError` happens](https://github.com/PostHog/posthog/pull/16473) to keep the workflow going, and handle [query output](https://github.com/PostHog/posthog/pull/16247) errors from Snowflake.
```

Comment:
>  define which destination errors are [retriable and which aren’t]

The rule of thumb is that if it's an error on the user's side (e.g. incorrect credentials, database being down/unreachable, incorrect schema names) then we do not retry. If it's an error on our side (e.g. pods take too long, transient network errors) then we retry as much as possible.